### PR TITLE
Gitpod prebuilds

### LINKED ISSUE #97: Status command to see if any repos are out-of-date
After the initial clone during setup, it may not be easy to fetch any upstream changes for the individual repos.

To start, maybe a ` ./lila-docker status` command can show if any of the local repos need to be updated.

### PR DESCRIPTION
**Pull request by Carbrex**
Also closes #97

After this pr, we can enable gitpod prebuilds for this repo and reduce the starting time of gitpod workspace to 3-4 minutes compared to 6-8 minutes without prebuilds on gitpod large workspace.

To try gitpod prebuilds follow the steps in gitpod-prebuilds.md but since it isn't merged follow steps 2-5 and put my fork `github.com/Carbrex/lila-docker` and `gitpod-prebuilds` branch on gitpod instead of your fork.

`fast-setup` command for opinionated setup.
`pull` and `status` command for status of github repos in repos/ dir.

### CODE REVIEW

File: lila-docker [hunk 1]
```diff
@@ -27,6 +27,44 @@ add_services() {
...
+    SETUP_DATABASE=true
+    setup_database
+    if is_gitpod_prebuild; then
+        touch ran_as_gitpod_prebuild
+        # wait for the services to be ready sleep for 5 seconds multiple times and check 8080 port if 502 status code is returned
+        for i in {1..100}; do # 100 * 5 = 500 seconds
+            sleep 5
+            status_code=$(curl -s -o /dev/null -w "%{http_code}" http://localhost:8080/)
+            echo "Status code: $status_code"
+            echo "Waited for $((i * 5)) seconds"
```
Comment (fitztrev) [inline lila-docker:60]: maybe consolidate the output of these lines. I wasn't sure at first what it was doing in the gitpod prebuild logs

from:
```
Waited for 200 seconds
Status code: 502
Waited for 205 seconds
Status code: 502
```

to:
```
Checking if lila is ready... not yet (502 after 200 seconds)
Checking if lila is ready... not yet (502 after 205 seconds)
```

File: pull-all [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            main_or_master="main"
+        elif [[ $(git branch --list master) ]]; then
+            main_or_master="master"
+        else
+            echo "$dir : doesn't have a main or master branch."
+            cd "$initial_dir"
+            continue
+        fi
+
+        git fetch upstream "$main_or_master" 2>/dev/null
```
[the remark this answers is not in the document]
Comment (Carbrex) [reply pull-all]: Should work now?

File: pull-all [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        git fetch "$upstream_or_origin" "$main_or_master" 2>/dev/null
+
+        # if repo is not on main or master branch, don't do anything
+        if [[ $(git branch --show-current) != "$main_or_master" ]]; then
+            echo "🔸 not on $main_or_master branch."
+            cd "$initial_dir"
+            continue
+        fi
+        # check if repo is up to date
+        if [[ $(git rev-list HEAD...upstream/$main_or_master --count) -eq 0 ]]; then
```
Comment (Carbrex) [inline pull-all:68]: I missed upstream here.

File: pull-all [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        # check if repo is up to date
+        if [[ $(git rev-list HEAD...$upstream_or_origin/$main_or_master --count) -eq 0 ]]; then
+            echo "✅ up to date"
+            cd "$initial_dir"
+            continue
+        fi
+
+        if [[ $pull_flag == true ]]; then
+            echo -n "pulling changes..."
+            git pull 2>/dev/null && echo "✅ done" || echo "❌ failed"
```
Comment (Carbrex) [inline pull-all:76]:
```suggestion
            git pull &>/dev/null && echo "✅ done" || echo "❌ failed"
```
Hide both stdout and stderr?

**Comment by fitztrev** [comment]
Pushed some possible changes to https://github.com/Carbrex/lila-docker/pull/3 for your review first

**Pull request merged**